src/tactic/has_variable_names.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2020 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jannis Limperg
+-/
+
+import tactic.core
+
+/-!
+# A tactic for type-based naming of variables
+
+When we name hypotheses or variables, we often do so in a type-directed fashion:
+a hypothesis of type `ℕ` is called `n` or `m`; a hypothesis of type `list ℕ` is
+called `ns`; etc. This module provides a tactic,
+`tactic.typical_variable_names`, which looks up typical variable names for a
+given type. Typical variable names are registered by giving an instance of the
+type class `has_variable_names`.
+-/
+
+universes u v
+
+/--
+Type class for associating a type `α` with typical variable names for elements
+of `α`. See `tactic.typical_variable_names`.
+-/
+meta class has_variable_names (α : Sort u) : Type :=
+(first_name : name)
+(other_names : list name)
+
+
+namespace tactic
+
+/--
+`typical_variable_names t` obtains typical names for variables of type `t`.
+The returned list is guaranteed to be nonempty. Fails if there is no instance
+`has_typical_variable_names t`.
+
+```
+typical_variable_names `(ℕ) = [`n, `m, `o]
+```
+-/
+meta def typical_variable_names (t : expr) : tactic (list name) :=
+(do
+  first_name ← to_expr ``(has_variable_names.first_name %%t),
+  n ← eval_expr name first_name,
+  other_names ← to_expr ``(has_variable_names.other_names %%t),
+  ns ← eval_expr (list name) other_names,
+  pure $ n :: ns)
+<|> fail! "typical_variable_names: unable to get typical variable names for type {t}"
+
+end tactic
+
+
+namespace has_variable_names
+
+/--
+`@make_listlike_instance α _ β` creates an instance `has_variable_names β` from
+an instance `has_variable_names α`. If `α` has associated names `a`, `b`, ...,
+the generated instance for `β` has names `as`, `bs`, ... This can be used to
+create instances for 'containers' such as lists or sets.
+-/
```

Comment:
Is this the reason that the tactic uses a typeclass rather than an annotation on the type constructor?

Reply:
Thank you for the super-quick review! I hadn't thought about using an annotation instead of a type class. But I think the type class approach is generally more flexible: people can add their own instances and even override the provided instances to implement whatever naming scheme they want.

Reply:
Could you mention these reasons somewhere in the module doc?

Reply:
Yes, this is a great idea. Will do.